library/std/src/sync/mutex.rs

```diff
@@ -399,6 +399,73 @@ impl<T: ?Sized> Mutex<T> {
         let data = self.data.get_mut();
         poison::map_result(self.poison.borrow(), |_| data)
     }
+
+    /// Run the given closure while holding the lock. The guarded value is mutably passed to the
+    /// closure, and the result of the closure is returned from `with`.
+    ///
+    /// This function is functionally equivalent to calling `lock` before the closure and dropping
+    /// the `MutexGuard` immediately after.
+    ///
+    /// # Errors
+    ///
+    /// If another user of this mutex panicked while holding the mutex, then
+    /// this call will return an error once the mutex is acquired.
+    ///
+    /// # Panics
+    ///
+    /// This function might panic when called if the lock is already held by
+    /// the current thread.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// #![feature(mutex_with)]
+    /// use std::sync::Mutex;
+    ///
+    /// let mutex = Mutex::new(0);
+    /// let result = mutex.with(|myint| { *myint += 1; *myint * 42 }).unwrap();
+    /// assert_eq!(*mutex.lock().unwrap(), 1);
+    /// assert_eq!(result, 42);
+    /// ```
+    #[unstable(feature = "mutex_with", issue = "none")]
+    pub fn with<R, F>(&self, f: F) -> Result<R, PoisonError<MutexGuard<'_, T>>>
```

Comment:
I *think* if you take `&mut self` instead, you can make otherwise deadlocking code to not compile:

```rust
let mutex = Mutex::new(0);

mutex.with(|_| {
    mutex.with(|_| {
        // this won't be reached anyway if I read the docs of lock correctly
    });
});
```

I just wanted to mention this because this is actually part of the embedded Mutex trait's [RFC](https://github.com/rust-embedded/wg/blob/master/rfcs/0377-mutex-trait.md#design-decisions-and-compatibility), but since one can cause deadlocks by locking twice anyway, I'm not sure how relevant my point is here.

Reply:
If you had a `&mut` to the `Mutex`, you have unique access to it and there's no need for locking at all. See also [`Mutex::get_mut`](https://doc.rust-lang.org/stable/std/sync/struct.Mutex.html#method.get_mut).

Reply:
I'm happy to added some sort of `with_mut` of people would find it useful, but I agree with @m-ou-se that the whole point of the PR is limiting the dynamic lifetime of a lock guard, not the static lifetime.

Reply:
Well, feel free to disregard my comment then :)